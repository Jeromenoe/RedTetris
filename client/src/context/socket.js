import React from 'react';
import socketio from "socket.io-client";


const localHost = "http://localhost:3001"; 
const herokuHost = "https://redtetris42-heroku.herokuapp.com"
const socketURL = process.env.NODE_ENV === "production" ? herokuHost : localHost

export const socket = socketio.connect(socketURL);
export const SocketContext = React.createContext();