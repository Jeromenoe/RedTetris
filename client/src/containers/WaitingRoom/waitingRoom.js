import { useSelector } from 'react-redux';
import React, { useContext } from 'react';
import { SocketContext } from '../../context/socket';
import { WaitingRoomContainer, WaitingRoomHeaderContainer, WaitingPlayer, ButtonContainer} from './styles'
import { GameSettings, showSettingPopup } from '../GameSettings/gameSettings'
import { getGameInfos, getPlayerDatas } from '../../selectors/game'
import store from '../../store/index'
import { updateData } from '../../actions/game';
import { initialState } from '../../reducers/index';

export const WaitingRoom = () => {
	const socket = useContext(SocketContext)
    const me = useSelector(getPlayerDatas);
    const game = useSelector(getGameInfos);
    const playerList = getPlayerList(game, me);
    const meIsAdmin = (game.players[0].name === me.name) ? true : false;

    const sendStartGameSignal = () => {
        const game = store.getState().game;
        socket.emit('setSettings', game);
		socket.emit('readyToPlay', null);
    };
    const sendPlayerLeaveSignal = () => {
        socket.emit('playerLeave', null);
        store.dispatch(updateData(initialState));
    };
    return (
        <WaitingRoomContainer>
            <WaitingRoomHeaderContainer>
                <h2 className="waitingRoomTitle">{game.name.toUpperCase()}</h2>
                { meIsAdmin && <button className="settingsBtn" id='showSettingPopup'><img src="./settingBtn.png" className="settingsImg" alt="settings" onClick={showSettingPopup}/></button> }
            </WaitingRoomHeaderContainer>
            <ul className="waitingPlayerList">{playerList}</ul>   
            <ButtonContainer>
                <button className="LeaveButton" onClick={sendPlayerLeaveSignal}>LEAVE</button>
                { meIsAdmin && <button className="PlayButton" onClick={sendStartGameSignal}>PLAY</button> }
                { !meIsAdmin && <button className="WaitingAdminButton">WAITING FOR {game.players[0].name.toUpperCase()}</button>}
            </ButtonContainer>
            <GameSettings/>
        </WaitingRoomContainer>
    )
}

function getPlayerList(game, me) {
    return game.players.map((player, index) => {
        var playerColor = null;
        if (player.name === me.name) {
            playerColor = "#f53643";
            return <WaitingPlayer color={playerColor} key={index}>{player.name + ' (me)'}</WaitingPlayer>
        }
        playerColor = "#5c5c5c";
        return <WaitingPlayer color={playerColor} key={index}>{player.name}</WaitingPlayer>
        });
}