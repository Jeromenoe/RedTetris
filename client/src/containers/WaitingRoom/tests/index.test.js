import React from 'react';
import {WaitingRoom} from '../waitingRoom';

import configureStore from "redux-mock-store";
import {Provider} from 'react-redux';
import toJson from "enzyme-to-json";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { mount, configure } from 'enzyme';

import { SocketContext, socket } from "../../../context/socket"


configure({adapter: new Adapter()});

beforeEach(() => {
	// Avoid `attachTo: document.body` Warning
	const div = document.createElement('div');
	div.setAttribute('id', 'container');
	document.body.appendChild(div);
  });
  
  afterEach(() => {
	const div = document.getElementById('container');
	if (div) {
	  document.body.removeChild(div);
	}
  });

describe('Testing WaitingRoom Container', () => {
	test('Check render', async () => {
		const mockStore = configureStore();
		const initialState = {
			game: {
				name: 'game',
				isStarted: false,
				isOver: false,
				opponents: [],
				players: [{name: 'toto'}],
				removeOwnIndestructibleLine: false,
				difficulty: 2,
				nbOfNextTetromino: 3,
			},
			me: {
				name: "toto",
				hasLost: false,
				stage: [[]],
				nextPieces: [],
				rows: 0,
				level: 0,
				score: 0
			}
		};
		const store = mockStore(initialState);
		
		const wrapper = mount(
		<Provider store={store}><SocketContext.Provider value={socket}><WaitingRoom  id="gameSettingsPopup" /></SocketContext.Provider></Provider>, 
		{ attachTo: document.getElementById('container') });
		expect(toJson(wrapper)).toMatchSnapshot();
		wrapper.find('#showSettingPopup').simulate('click');
	  });
});