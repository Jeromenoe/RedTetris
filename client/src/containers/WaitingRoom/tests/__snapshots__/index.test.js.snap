// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing WaitingRoom Container Check render 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <WaitingRoom
    id="gameSettingsPopup"
  >
    <styled.div>
      <div
        className="sc-bdnxRM kNCjGk"
      >
        <styled.div>
          <div
            className="sc-dlnjwi kYnDVn"
          >
            <h2
              className="waitingRoomTitle"
            >
              GAME
            </h2>
            <button
              className="settingsBtn"
              id="showSettingPopup"
            >
              <img
                alt="settings"
                className="settingsImg"
                onClick={[Function]}
                src="./settingBtn.png"
              />
            </button>
          </div>
        </styled.div>
        <ul
          className="waitingPlayerList"
        >
          <styled.li
            color="#f53643"
            key="0"
          >
            <li
              className="sc-hKFxyN eTyhPE"
              color="#f53643"
            >
              toto (me)
            </li>
          </styled.li>
        </ul>
        <styled.div>
          <div
            className="sc-gtsrHT hVRCUb"
          >
            <button
              className="LeaveButton"
              onClick={[Function]}
            >
              LEAVE
            </button>
            <button
              className="PlayButton"
              onClick={[Function]}
            >
              PLAY
            </button>
          </div>
        </styled.div>
        <GameSettings>
          <styled.div
            id="gameSettingsPopup"
          >
            <div
              className="sc-eCApnc dxdVpO"
              id="gameSettingsPopup"
            >
              <styled.div>
                <div
                  className="sc-jSFjdj eXbZAp"
                >
                  <div
                    className="settings-header"
                  >
                    <span
                      className="CloseButton"
                      id="closebtn"
                      onClick={[Function]}
                    >
                      Ã—
                    </span>
                    <h2>
                      SETTINGS
                    </h2>
                  </div>
                  <ul
                    className="options-list"
                  >
                    <li
                      className="options-elem"
                      key="1"
                    >
                      DIFFICULTY:
                      <RadioOptions
                        settingId={1}
                      >
                        <div
                          className="radio-group"
                        >
                          <input
                            checked={false}
                            id="opt-1"
                            name="selector-1"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-1"
                          >
                            EASY
                          </label>
                          <input
                            checked={true}
                            id="opt-2"
                            name="selector-1"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-2"
                          >
                            MEDIUM
                          </label>
                          <input
                            checked={false}
                            id="opt-3"
                            name="selector-1"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-3"
                          >
                            HARD
                          </label>
                          <input
                            checked={false}
                            id="opt-4"
                            name="selector-1"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-4"
                          >
                            LEVEL
                          </label>
                        </div>
                      </RadioOptions>
                    </li>
                    <li
                      className="options-elem"
                      key="2"
                    >
                      NUMBER OF NEXT TETROMINOS:
                      <RadioOptions
                        settingId={2}
                      >
                        <div
                          className="radio-group"
                        >
                          <input
                            checked={false}
                            id="opt-5"
                            name="selector-2"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-5"
                          >
                            1
                          </label>
                          <input
                            checked={false}
                            id="opt-6"
                            name="selector-2"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-6"
                          >
                            2
                          </label>
                          <input
                            checked={true}
                            id="opt-7"
                            name="selector-2"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-7"
                          >
                            3
                          </label>
                          <input
                            checked={false}
                            id="opt-8"
                            name="selector-2"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-8"
                          >
                            4
                          </label>
                          <input
                            checked={false}
                            id="opt-9"
                            name="selector-2"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-9"
                          >
                            5
                          </label>
                        </div>
                      </RadioOptions>
                    </li>
                    <li
                      className="options-elem"
                      key="3"
                    >
                      DELETE OWN INDESTRUTIBLE LINES:
                      <RadioOptions
                        settingId={3}
                      >
                        <div
                          className="radio-group"
                        >
                          <input
                            checked={true}
                            id="opt-10"
                            name="selector-3"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-10"
                          >
                            FALSE
                          </label>
                          <input
                            checked={false}
                            id="opt-11"
                            name="selector-3"
                            onChange={[Function]}
                            type="radio"
                          />
                          <label
                            htmlFor="opt-11"
                          >
                            TRUE
                          </label>
                        </div>
                      </RadioOptions>
                    </li>
                  </ul>
                </div>
              </styled.div>
            </div>
          </styled.div>
        </GameSettings>
      </div>
    </styled.div>
  </WaitingRoom>
</Provider>
`;
